Backpack/src/main/java/net/skyscanner/backpack/calendar2/adapter/CalendarAdapter.kt

```diff
@@ -0,0 +1,98 @@
+/**
+ * Backpack for Android - Skyscanner's Design System
+ *
+ * Copyright 2018-2021 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.skyscanner.backpack.calendar2.adapter
+
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import net.skyscanner.backpack.calendar2.CalendarFooterAdapter
+import net.skyscanner.backpack.calendar2.CalendarState
+import net.skyscanner.backpack.calendar2.data.CalendarDay
+import net.skyscanner.backpack.calendar2.data.CalendarFooter
+import net.skyscanner.backpack.calendar2.data.CalendarHeader
+import net.skyscanner.backpack.calendar2.data.CalendarItem
+import net.skyscanner.backpack.calendar2.data.CalendarMonth
+import net.skyscanner.backpack.calendar2.data.CalendarSpace
+import net.skyscanner.backpack.calendar2.extension.getItem
+import net.skyscanner.backpack.calendar2.extension.yearMonthHash
+import net.skyscanner.backpack.util.Consumer
+import net.skyscanner.backpack.util.ItemHolder
+import org.threeten.bp.temporal.ChronoField
+
+internal class CalendarAdapter(
+  scope: CoroutineScope,
+  input: StateFlow<CalendarState>,
+  private val footers: CalendarFooterAdapter<*>,
+  private val output: Consumer<CalendarDay>,
+) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+
+  private var data: List<CalendarMonth> = input.value.months
+
+  init {
+    setHasStableIds(true)
+    input.onEach {
+      this.data = it.months
+      notifyDataSetChanged()
```

Comment:
I might be wrong, but not sure if that would make much difference as we're using `setHasStableIds` on this adapter already

Reply:
yes I think you're right - it's been a long time since I used stable ids - afaik stable ids were mainly introduced to make it easier to migrate from `ListView` - I remember Yigit (who did a lot of the original `RecyclerView` work) mentioned that he regrets adding support for it - using diffutil is the recommended approach (and would also remove the need for `YearMonth`